toolchain/check/check.cpp

```diff
@@ -56,16 +85,218 @@ auto CheckParseTree(SharedValueStores& value_stores,
 
   context.VerifyOnFinish();
 
-  sem_ir.set_has_errors(err_tracker.seen_error());
+  sem_ir.set_has_errors(unit_info.err_tracker.seen_error());
 
 #ifndef NDEBUG
   if (auto verify = sem_ir.Verify(); !verify.ok()) {
     CARBON_FATAL() << sem_ir << "Built invalid semantics IR: " << verify.error()
                    << "\n";
   }
 #endif
+}
+
+// Returns a key form of the package object.
+static auto PackageKey(UnitInfo& unit_info, Parse::Tree::Package package)
+    -> std::pair<llvm::StringRef, llvm::StringRef> {
+  auto* stores = unit_info.unit->value_stores;
+  return {package.package_id.is_valid()
+              ? stores->identifiers().Get(package.package_id)
+              : "",
+          package.library_id.is_valid()
+              ? stores->string_literals().Get(package.library_id)
+              : ""};
+}
+
+static constexpr llvm::StringLiteral ExplicitMainName = "Main";
 
-  return sem_ir;
+// Marks an import as required on both the source and target file. seen_imports
+// is optional, and will be omitted for implicit imports (which are disallowed
+// when explicit).
+static auto TrackImport(
+    llvm::DenseMap<std::pair<llvm::StringRef, llvm::StringRef>, UnitInfo*>&
+        api_map,
+    llvm::DenseMap<std::pair<llvm::StringRef, llvm::StringRef>, Parse::Node>*
+        explicit_import_map,
+    UnitInfo& unit_info, Parse::Tree::Package import) -> void {
+  auto package_key = PackageKey(unit_info, import);
+
+  // Specialize the error for imports from `Main`.
+  if (package_key.first == ExplicitMainName) {
+    // Implicit imports will have already warned.
+    if (explicit_import_map) {
+      CARBON_DIAGNOSTIC(ImportMainPackage, Error,
+                        "The `Main` package cannot be mentioned explicitly on "
+                        "imports; it can only be imported implicitly");
```

Comment:
> I think it'd be useful to split this diagnostic into two:
> 
> * If the current library and the imported library are in the same package: "Use `import library` instead of `import package PackageName library` to import from the same package." or something like that.

I've added a TODO for this, as I'm already working on the implicit package name support. I'd like to test it together with that.

> * If the current package is not `Main`: "Cannot import `Main` package from another package."

I've rephrased the warning.

> It looks like the first of those two rules isn't being enforced yet.

I'm not sure it's splitting when one isn't handled. ;)

Reply:
> I'm not sure it's splitting when one isn't handled. ;)

You are currently enforcing the first rule for the `Main` package, but not for any other package, which is what I was suggesting you could split. The diagnostic we currently get from importing a `Main` library from another `Main` library  is incorrect: "Cannot import `Main` from other packages."

But no action required here: I'm fine with this as-is given the TODO you added.